cmd/ipfs/util/signal_windows.go

```diff
@@ -0,0 +1,62 @@
+package util
+
+import (
+	"os"
+
+	"golang.org/x/sys/windows"
+)
+
+type (
+	DWORD    = uint32
+	MSVCBOOL = uintptr
+)
+
+const (
+	processed = iota
+	notProcessed
+)
+
+func init() {
+	nativeNotifier = windowEventHandler
+}
+
+// windowEventHandler creates and registers a WINAPI `HandlerRoutine` callback
+// which receives window events and relays them as interrupt signals to the provided channel
+func windowEventHandler(ih *IntrHandler, sigChan chan os.Signal, sigs ...os.Signal) {
+	// we don't want to relay the events `signal` is already listening for
+	// otherwise the channel will receive them twice
+	var goIsHandlingIntterupts bool
+	for _, sig := range sigs {
+		if sig == os.Interrupt {
+			goIsHandlingIntterupts = true
+		}
+	}
+
+	fp := func(ctrlType DWORD) MSVCBOOL {
+		switch ctrlType {
+		case windows.CTRL_CLOSE_EVENT:
+			//NOTE: the OS will terminate our process after receiving this event
+			// either after `SPI_GETHUNGAPPTIMEOUT` or immediately after we return non-0
+			// (whichever is first)
+			sigChan <- os.Interrupt
+			ih.wg.Wait()
+			return processed
+		case windows.CTRL_C_EVENT, windows.CTRL_BREAK_EVENT:
+			if !goIsHandlingIntterupts {
+				sigChan <- os.Interrupt
+			}
+			return processed
```

Comment:
If go is handling interrupts, will this cause problems? I wonder if we should just _always_ use this function on windows (skipping the `signal.Notify` call).